File: src/main/java/me/aragot/hglmoderation/tools/PlayerUtils.java
package me.aragot.hglmoderation.tools;

import org.bson.Document;

import javax.net.ssl.HttpsURLConnection;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.HashMap;
import java.util.Objects;

public class PlayerUtils {

    private static final HashMap<String, String> playerNameCache = new HashMap<>();

    public static String getUsernameFromUUID(String uuid){
        if(playerNameCache.containsKey(uuid))
            return playerNameCache.get(uuid);

        try {
            URL url = new URL("https://sessionserver.mojang.com/session/minecraft/profile/" + uuid);
            HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
            connection.setRequestProperty("Content-Type", "application/json");
            connection.setConnectTimeout(2000);
            connection.setReadTimeout(2000);

            int status = connection.getResponseCode();

            if(status > 299) {
                return null;
            }

            BufferedReader in = new BufferedReader(new InputStreamReader(connection.getInputStream()));

            StringBuilder content = new StringBuilder();
            String inputLine;
            while ((inputLine = in.readLine()) != null) {
                content.append(inputLine);
            }

            in.close();
            connection.disconnect();
            Document doc = Document.parse(content.toString());
            playerNameCache.put(uuid, doc.getString("name"));
            return doc.getString("name");

        } catch (IOException e) {
            return "null";
        }
    }

    public static String getUuidFromUsername(String username){
        if(playerNameCache.containsKey(username))
            return playerNameCache.values().stream()
                    .filter(name -> Objects.equals(name, username))
                    .findFirst().orElse("None");

        try {
            URL url = new URL("https://api.mojang.com/users/profiles/minecraft/" + username);
            HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
            connection.setRequestProperty("Content-Type", "application/json");
            connection.setConnectTimeout(2000);
            connection.setReadTimeout(2000);

            int status = connection.getResponseCode();

            if(status > 299) {
                return null;
            }

            BufferedReader in = new BufferedReader(new InputStreamReader(connection.getInputStream()));

            StringBuilder content = new StringBuilder();
            String inputLine;
            while ((inputLine = in.readLine()) != null) {
                content.append(inputLine);
            }

            in.close();
            connection.disconnect();
            Document doc = Document.parse(content.toString());

            return addHyphensToUUID(doc.getString("id"));

        } catch (IOException e) {
            return null;
        }
    }

    public static String addHyphensToUUID(String uuidWithoutHyphens) {
        if (uuidWithoutHyphens.length() != 32) {
            throw new IllegalArgumentException("Invalid UUID length");
        }

        StringBuilder sb = new StringBuilder(uuidWithoutHyphens);
        sb.insert(20, "-");
        sb.insert(16, "-");
        sb.insert(12, "-");
        sb.insert(8, "-");

        return sb.toString();
    }

    public static void removePlayerFromCache(String uuid){
        playerNameCache.remove(uuid);
    }
}

File: src/main/java/me/aragot/hglmoderation/service/player/PlayerUtils.kt
package me.aragot.hglmoderation.service.player

import com.velocitypowered.api.util.UuidUtils
import me.aragot.hglmoderation.HGLModeration
import org.bson.Document
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.URL
import java.util.*
import javax.net.ssl.HttpsURLConnection
import kotlin.collections.ArrayList

class PlayerUtils {
    companion object {
        private val playerCache = ArrayList<PlayerCacheEntry>()

        fun getUsernameFromUUID(uuid: UUID): String? {
            val cacheEntry = this.getCacheEntry(uuid = uuid)
            return if (cacheEntry !== null && cacheEntry.isValid()) {
                cacheEntry.username
            } else {
                try {
                    val url = URL("https://sessionserver.mojang.com/session/minecraft/profile/$uuid")
                    val connection = url.openConnection() as HttpsURLConnection
                    connection.setRequestProperty("Content-Type", "application/json")
                    connection.connectTimeout = 2000
                    connection.readTimeout = 2000
                    val status = connection.responseCode

                    if (status > 299) {
                        return null
                    }

                    val reader = BufferedReader(InputStreamReader(connection.inputStream))
                    val content = StringBuilder()
                    var inputLine: String?
                    while (reader.readLine().also { inputLine = it } != null) {
                        content.append(inputLine)
                    }
                    reader.close()
                    connection.disconnect()
                    val doc = Document.parse(content.toString())
                    val name = doc.getString("name")
                    if (cacheEntry !== null) {
                        cacheEntry.username = name
                        cacheEntry.updateTimeStamp()
                    } else {
                        playerCache.add(PlayerCacheEntry(name, uuid))
                    }

                    name
                } catch (e: IOException) {
                    "null"
                }
            }
        }

        fun getUuidFromUsername(username: String): UUID? {
            val cacheEntry = this.getCacheEntry(username = username)
            return if (cacheEntry !== null && cacheEntry.isValid()) cacheEntry.uuid
            else try {
                val apiLink = "https://api.mojang.com/users/profiles/minecraft/$username"
                val url = URL(apiLink)
                val connection = url.openConnection() as HttpsURLConnection
                connection.setRequestProperty("Content-Type", "application/json")
                connection.connectTimeout = 2000
                connection.readTimeout = 2000
                val status = connection.responseCode
                HGLModeration.instance.logger.info("Fetched $apiLink and got response code: $status")
                if (status > 299) {
                    return null
                }

                val reader = BufferedReader(InputStreamReader(connection.inputStream))
                val content = StringBuilder()
                var inputLine: String?
                while (reader.readLine().also { inputLine = it } != null) {
                    content.append(inputLine)
                }
                reader.close()
                connection.disconnect()

                val doc = Document.parse(content.toString())
                val uuid = UuidUtils.fromUndashed(doc.getString("id"))
                if (cacheEntry !== null) {
                    cacheEntry.uuid = uuid
                    cacheEntry.updateTimeStamp()
                } else {
                    playerCache.add(PlayerCacheEntry(username, uuid))
                }

                uuid
            } catch (e: IOException) {
                null
            }
        }

        fun removePlayerFromCache(uuid: UUID? = null, username: String? = null) {
            if (uuid === null && username === null) return
            playerCache.removeIf { entry: PlayerCacheEntry -> (username === entry.username || uuid === entry.uuid) }
        }

        private fun getCacheEntry(username: String? = null, uuid: UUID? = null): PlayerCacheEntry? {
            return if (username === null && uuid === null) null
            else playerCache.find { entry: PlayerCacheEntry -> (username.equals(entry.username, ignoreCase = true) || uuid === entry.uuid) }
        }
    }
}